File: lib/to_mdx/page_dimensions.rb
# frozen_string_literal: true

require_relative "./parsers/open_document/base"

module ToMdx
  module PageDimensions
    class << self
      def infer(content_doc, styles_doc)
        infer_from_master_page(content_doc, styles_doc) ||
          infer_from_page_layout(styles_doc) ||
          largest_page_frame(content_doc)
      end

      private

      def infer_from_master_page(content_doc, styles_doc)
        master_page_counts = {}
        content_doc.elements.each("//draw:page") do |page|
          master_page = page.attributes["draw:master-page-name"]
          if master_page
            master_page_counts[master_page] ||= 0
            master_page_counts[master_page] += 1
          end
        end

        most_common_master_page = master_page_counts.max_by { |_, count| count }&.first
        return nil unless most_common_master_page

        master_page = REXML::XPath.first(styles_doc, "//style:master-page/@style:name[.='#{most_common_master_page}']")&.element
        return nil unless master_page

        page_layout_name = master_page.attributes["style:page-layout-name"]
        return nil unless page_layout_name

        page_layout = REXML::XPath.first(styles_doc, "//style:page-layout/@style:name[.='#{page_layout_name}']")&.element
        return nil unless page_layout

        props = page_layout.elements["style:page-layout-properties"]
        return nil unless props

        width_attr = props.attributes["fo:page-width"]
        height_attr = props.attributes["fo:page-height"]
        return nil unless width_attr && height_attr

        width_cm = extract_cm_value(width_attr)
        height_cm = extract_cm_value(height_attr)
        return { width: width_cm, height: height_cm } if width_cm && height_cm

        nil
      end

      def infer_from_page_layout(styles_doc)
        styles_doc.elements.each("//style:page-layout-properties") do |props|
          width_attr = props.attributes["fo:page-width"]
          height_attr = props.attributes["fo:page-height"]

          next unless width_attr && height_attr

          width_cm = extract_cm_value(width_attr)
          height_cm = extract_cm_value(height_attr)

          next unless width_cm && height_cm

          return { width: width_cm, height: height_cm }
        end

        nil
      end

      def largest_page_frame(content_doc)
        max_width = 0
        max_height = 0

        content_doc.elements.each("//draw:frame") do |frame|
          width_attr = frame.attributes["svg:width"]
          height_attr = frame.attributes["svg:height"]

          next unless width_attr && height_attr

          width_cm = extract_cm_value(width_attr)
          height_cm = extract_cm_value(height_attr)

          next unless width_cm && height_cm

          if width_cm > max_width
            max_width = width_cm
          end

          if height_cm > max_height
            max_height = height_cm
          end
        end

        return nil if max_width == 0 || max_height == 0

        { width: max_width, height: max_height }
      end

      def extract_cm_value(value)
        match = value.match(/^(-?[\d.]+)cm$/)
        return nil unless match

        match[1].to_f
      end
    end
  end
end

File: src/page-dimensions.ts
export interface PageDimensions {
  width: number;
  height: number;
}

export function infer(contentDoc: Document, stylesDoc: Document): PageDimensions | null {
  return inferFromMasterPage(contentDoc, stylesDoc) || inferFromPageLayout(stylesDoc) || largestPageFrame(contentDoc);
}

function inferFromMasterPage(contentDoc: Document, stylesDoc: Document): PageDimensions | null {
  const masterPageCounts: Record<string, number> = {};

  // Find most common master page
  const pages = contentDoc.getElementsByTagName("draw:page");
  for (let i = 0; i < pages.length; i++) {
    const page = pages[i] as Element;
    const masterPage = page.getAttribute("draw:master-page-name");
    if (masterPage) {
      masterPageCounts[masterPage] = (masterPageCounts[masterPage] || 0) + 1;
    }
  }

  const mostCommonMasterPage = Object.entries(masterPageCounts).reduce(
    (max, [name, count]) => (count > max.count ? { name, count } : max),
    { name: "", count: 0 },
  ).name;

  if (!mostCommonMasterPage) return null;

  // Find the master page element
  const masterPages = stylesDoc.getElementsByTagName("style:master-page");
  let masterPageElement: Element | null = null;

  for (let i = 0; i < masterPages.length; i++) {
    const mp = masterPages[i] as Element;
    if (mp.getAttribute("style:name") === mostCommonMasterPage) {
      masterPageElement = mp;
      break;
    }
  }

  if (!masterPageElement) return null;

  const pageLayoutName = masterPageElement.getAttribute("style:page-layout-name");
  if (!pageLayoutName) return null;

  // Find the page layout
  const pageLayouts = stylesDoc.getElementsByTagName("style:page-layout");
  let pageLayout: Element | null = null;

  for (let i = 0; i < pageLayouts.length; i++) {
    const pl = pageLayouts[i] as Element;
    if (pl.getAttribute("style:name") === pageLayoutName) {
      pageLayout = pl;
      break;
    }
  }

  if (!pageLayout) return null;

  const props = pageLayout.getElementsByTagName("style:page-layout-properties")[0];
  if (!props) return null;

  const widthAttr = props.getAttribute("fo:page-width");
  const heightAttr = props.getAttribute("fo:page-height");

  if (!widthAttr || !heightAttr) return null;

  const width = extractCmValue(widthAttr);
  const height = extractCmValue(heightAttr);

  return width !== null && height !== null ? { width, height } : null;
}

function inferFromPageLayout(stylesDoc: Document): PageDimensions | null {
  const pageLayoutProps = stylesDoc.getElementsByTagName("style:page-layout-properties");

  for (let i = 0; i < pageLayoutProps.length; i++) {
    const props = pageLayoutProps[i];
    const widthAttr = props.getAttribute("fo:page-width");
    const heightAttr = props.getAttribute("fo:page-height");

    if (!widthAttr || !heightAttr) continue;

    const width = extractCmValue(widthAttr);
    const height = extractCmValue(heightAttr);

    if (width !== null && height !== null) {
      return { width, height };
    }
  }

  return null;
}

function largestPageFrame(contentDoc: Document): PageDimensions | null {
  let maxWidth = 0;
  let maxHeight = 0;

  const frames = contentDoc.getElementsByTagName("draw:frame");

  for (let i = 0; i < frames.length; i++) {
    const frame = frames[i];
    const widthAttr = frame.getAttribute("svg:width");
    const heightAttr = frame.getAttribute("svg:height");

    if (!widthAttr || !heightAttr) continue;

    const width = extractCmValue(widthAttr);
    const height = extractCmValue(heightAttr);

    if (width !== null && height !== null) {
      if (width > maxWidth) maxWidth = width;
      if (height > maxHeight) maxHeight = height;
    }
  }

  return maxWidth > 0 && maxHeight > 0 ? { width: maxWidth, height: maxHeight } : null;
}

function extractCmValue(value: string): number | null {
  const match = value.match(/^(-?[\d.]+)cm$/);
  return match ? parseFloat(match[1]) : null;
}
